### PR TITLE
Issue 2

### PR DESCRIPTION
**Pull request by mbarekh**
Replace dom parser with xml2js

### CODE REVIEW

File: viewer/src/index.ts [hunk 1]
```diff
@@ -6,6 +6,10 @@ const component = {
   graphical: document.currentScript.getAttribute("graphical")
 };
 const parser = new Parser(component);
-parser.parse();
-const drawComponent = new DrawComponent();
-drawComponent.draw(parser, "diagram");
\ No newline at end of file
+parser.parse((err, p) => {
+  if (err) {
+    return;
```
Comment (maurelio1234) [inline viewer/src/index.ts]: log error?

Comment (maurelio1234) [inline viewer/src/index.ts]: or display message?

File: viewer/src/parser.ts [hunk 2]
```diff
@@ -24,58 +27,66 @@ export class Parser {
...
-
-    private parseGraphical(): void {
+    private parseGraphical(graphicalJson): void {
         this.locations = {};
-        const xmlGraphicalDom = (new DOMParser()).parseFromString(this.componentGraphicalModel.graphical, "text/xml");
-        let stateGraphicalData = xmlGraphicalDom.getElementsByTagName(graphicalTags.States)[0].getElementsByTagName(graphicalTags.StateGraphicalData);
+        let stateGraphicalData = graphicalJson.ComponentViewModelGraphicalData.States[0].StateGraphicalData;
         for (let i = 0; i < stateGraphicalData.length; i++) {
-            const id = stateGraphicalData[i].getAttribute(graphicalTags.Id);
+            const id = stateGraphicalData[i].$.Id;
```
Comment (maurelio1234) [inline viewer/src/parser.ts]: curiosity, what's the meaning of "$" here?

Comment (maurelio1234) [inline viewer/src/parser.ts]: the original code was clearer

File: viewer/src/parser.ts [hunk 3]
```diff
@@ -253,20 +270,20 @@ export class Parser {
         }
     }
 
-    private setStates(scxmlDom: Document): void {
-        const statesDom = scxmlDom.getElementsByTagName(modelTags.StateData);
+    private setStates(modelJson): void {
```
Comment (maurelio1234) [inline viewer/src/parser.ts]: add type

File: viewer/src/parser.ts [hunk 4]
```diff
@@ -197,33 +215,33 @@ export class Parser {
         }
     }
 
-    private setLinks(scxmlDom: Document): void {
-        const linksDom = scxmlDom.getElementsByTagName(modelTags.TransitionData);
+    private setLinks(modelJson): void {
```
Comment (maurelio1234) [inline viewer/src/parser.ts]: type

File: viewer/src/parser.ts [hunk 5]
```diff
@@ -24,58 +27,66 @@ export class Parser {
         this.componentGraphicalModel = componentGraphicalModel;
     }
 
-    parse() {
-        this.parseGraphical();
-        this.parseModel();
-    }
-
-    private parseGraphical(): void {
+    private parseGraphical(graphicalJson): void {
```
Comment (maurelio1234) [inline viewer/src/parser.ts]: add typing

File: viewer/src/parser.ts [hunk 6]
```diff
@@ -24,58 +27,66 @@ export class Parser {
...
+            thisObject.parseModel(result);
+            parseListener(null, thisObject);
+        }).catch((err) => {
+            console.error("Parsing error");
+            console.error(err);
+            parseListener(err, null);
+        });
+    }
+
+    private parseModel(modelJson): void {
```
Comment (maurelio1234) [inline viewer/src/parser.ts]: add type

File: viewer/src/parser.ts [hunk 7]
```diff
@@ -24,58 +27,66 @@ export class Parser {
...
         this.finalStates = this.setFinalStates();
         this.entryPointState = this.getEntyPointState();
         this.nodeDataArray = this.setNodeDataArray();
         this.nodeDataArray = this.nodeDataArray.concat(this.linksLabel);
         this.addControlPoint();
     }
 
-    private setComponentName(scxmlDom: Document): void {
-        this.componentName = scxmlDom.getElementsByTagName(modelTags.ComponentViewModelData)[0].getAttribute(modelTags.Name);
+    private setComponentName(modelJson): void {
```
Comment (maurelio1234) [inline viewer/src/parser.ts]: type

File: viewer/src/parser.ts [hunk 8]
```diff
@@ -147,29 +159,35 @@ export class Parser {
         return nodeDataArray;
     }
 
-    private getControlPointTransition(transitionGraphicalData: NodeListOf<Element>): { [key: string]: Curve } {
+
+    private getControlPointTransition(transitionGraphicalData: Array<any>): { [key: string]: Curve } {
```
Comment (maurelio1234) [inline viewer/src/parser.ts]: any => type

File: viewer/src/parser.ts [hunk 9]
```diff
@@ -305,17 +322,17 @@ export class Parser {
         this.entryPointStateMachine = entryPointStateMachine;
     }
 
-    private setStateMachines(scxmlDom: Document): void {
-        const stateMachineDom = scxmlDom.getElementsByTagName(modelTags.StateMachineData);
+    private setStateMachines(modelJson): void {
```
Comment (maurelio1234) [inline viewer/src/parser.ts]: type

File: viewer/src/graphicalTypes.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    TransitionPatternStates: Array<TransitionPatternStates>;
+    TransversalLinks: Array<TransversalLinks>;
+}
+
+export interface Links {
+    TransitionGraphicalData: Array<TransitionGraphicalDataElement>;
+}
+
+export interface TransitionGraphicalDataElement {
+    $: Properties;
```
Comment (nicolaserny) [inline viewer/src/graphicalTypes.ts]: Why do you need to define a $ member ? Not very clear
$ is mainly by jquery stuffs..

File: viewer/src/graphicalTypes.ts [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    Points: Array<TransitionGraphicalDataPointElement>;
+    IsNameDisplayed: string;
+}
+
+export interface TransitionGraphicalDataPointElement {
+    Point: Array<PointGraphical>;
+}
+
+export interface PointGraphical {
+    X: string;
```
Comment (nicolaserny) [inline viewer/src/graphicalTypes.ts]: X and Y should be numbers ?

File: viewer/src/graphicalTypes.ts [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    Y: string;
+}
+
+export interface States {
+    StateGraphicalData: Array<$<StateGraphicalDataElement>>;
+}
+
+export interface StateGraphicalDataElement {
+    Id: string;
+    CenterX: string;
```
Comment (nicolaserny) [inline viewer/src/graphicalTypes.ts]: Same comment about numbers

File: viewer/src/graphicalTypes.ts [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    TransitionGraphicalData: Array<TransitionGraphicalDataElement>;
+}
+
+export interface Properties {
+    Color: string;
+    Curviness: string;
+    Id: string;
+}
+
+export interface $<T> {
```
Comment (nicolaserny) [inline viewer/src/graphicalTypes.ts]: What's the meaning of this interface ?

Comment (nicolaserny) [inline viewer/src/graphicalTypes.ts]: I think it's a variable from npm2js lib. But I think you can customise this behaviour

File: viewer/src/modelTypes.ts [hunk 14]
```diff
@@ -0,0 +1,93 @@
+export interface Model {
+    ComponentViewModelData: ComponentViewModelData;
+}
+
+export interface ComponentViewModelData {
+    ComponentViewModelData: ComponentViewModelData;
+    $: ModelProperties;
```
Comment (nicolaserny) [inline viewer/src/modelTypes.ts]: same comment about $

File: viewer/src/parser.ts [hunk 15]
```diff
@@ -19,20 +21,36 @@ export class Parser {
...
     private nodeDataArray: Array<NodeDataArrayTemplate>;
 
+    public promisifyParserString: any;
+
     constructor(componentGraphicalModel: ComponentGraphicalModel) {
         this.componentGraphicalModel = componentGraphicalModel;
+        this.promisifyParserString = promisify(parseString);
+    }
+
+    parse(parseListener: (error: Error, parser: Parser) => void) {
```
Comment (nicolaserny) [inline viewer/src/parser.ts]: Why do you need to have parser as parameter of your callback? Why not just the result of the parsing ? no?
Perhaps returning a promise could be a better choice than using callbacks

File: viewer/src/parser.ts [hunk 16]
```diff
@@ -19,20 +21,36 @@ export class Parser {
     private finalStates: Array<String>;
     private entryPointState: string;
     private entryPointStateMachine: string;
-    private stateMachineNames: Array<string>;
     private linkDataArray: Array<LinkDataArrayTemplate>;
     private nodeDataArray: Array<NodeDataArrayTemplate>;
 
+    public promisifyParserString: any;
```
Comment (nicolaserny) [inline viewer/src/parser.ts]: Why public ? should be private ? you don't want to expose how you parse

File: viewer/src/parser.ts [hunk 17]
```diff
@@ -2,8 +2,10 @@ import { graphicalTags, modelTags } from "configurationParser";
 import { LinkLabelTemplate, TransitionTemplate, TriggerableTransitionTemplate, StateMachineTemplate, StateTemplate, LinkDataArrayTemplate, NodeDataArrayTemplate } from "gojsTemplates";
 import { Point, Curve, StateMachine, State, ComponentGraphicalModel } from "parserObjects";
 import { finalStateColor, stateColor, transitionPatternStateColor, entryPointStateColor } from "graphicColors";
+import { Graphical, $, StateGraphicalDataElement, TransitionGraphicalDataElement } from "graphicalTypes";
 import { parseString } from "xml2js";
 import * as promisify from "es6-promisify";
+import { Model } from "modelTypes";
 
 export class Parser {
```
Comment (nicolaserny) [inline viewer/src/parser.ts]: Bad naming

File: viewer/test/parser.test.ts [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+    let componentGraphicalModel, model, graphical;
+    beforeEach(() => {
+        model = fs.readFileSync("./test/WorldHello.cxml").toString();
+        graphical = fs.readFileSync("./test/WorldHello_Graphical.xml").toString();
+        parser = new Parser({ model, graphical });
+        parser.promisifyParserString()
+    });
+
+    test("Test getLocationState method: should get the right locations of the state", (done) => {
+        parser.promisifyParserString(graphical).then((graphicalJson) => {
```
Comment (nicolaserny) [inline viewer/test/parser.test.ts]: I don't understand the way it should work. It's so complicated!
What I was expecting: parser.parse(...).then(...)

Comment (mbarekh) [inline viewer/test/parser.test.ts]: done

File: viewer/src/parser.ts (hunk 15 above)
Comment (mbarekh) [inline viewer/src/parser.ts]: done

File: viewer/src/parser.ts (hunk 16 above)
Comment (mbarekh) [inline viewer/src/parser.ts]: done

File: viewer/src/parser.ts (hunk 17 above)
Comment (mbarekh) [inline viewer/src/parser.ts]: done

File: viewer/src/index.ts [hunk 19]
```diff
@@ -1,16 +1,17 @@
...
 const component = {
   model: document.currentScript.getAttribute("model"),
   graphical: document.currentScript.getAttribute("graphical")
 };
-const parser = new Parser(component);
-parser.parse((err, p) => {
-  if (err) {
+const parser = new ComponentModelParser(component);
+parser.parse()
+  .then(() => {
```
Comment (nicolaserny) [inline viewer/src/index.ts]: Not a logical way to use the parser. It seems weird to put the parser as a parameter of drawComponent.
I think a better solution is: parser.parse().then((data) => ... drawComponent.draw(data...))
@maurelio1234 what do you think?

Comment (mbarekh) [inline viewer/src/index.ts]: done

File: viewer/src/gojsTemplates.ts [hunk 20]
```diff
@@ -49,4 +49,4 @@ export type NodeDataArrayTemplate = StateMachineTemplate | StateTemplate | LinkL
 export interface DrawComponentData {
     nodeDataArray: Array<NodeDataArrayTemplate>;
```
Comment (nicolaserny) [inline viewer/src/gojsTemplates.ts]: nodeDataArray is a redundant name (with class name and member type)
nodes could be a better choice.

File: viewer/src/gojsTemplates.ts [hunk 21]
```diff
@@ -49,4 +49,4 @@ export type NodeDataArrayTemplate = StateMachineTemplate | StateTemplate | LinkL
...
     linkDataArray: Array<NodeDataArrayTemplate>;
```
Comment (nicolaserny) [inline viewer/src/gojsTemplates.ts]: Same comment about naming

File: viewer/src/gojsTemplates.ts (hunk 20 above)
Comment (mbarekh) [inline viewer/src/gojsTemplates.ts]: gojs documentation uses this naming http://gojs.net/latest/intro/groups.html

File: viewer/src/gojsTemplates.ts (hunk 21 above)
Comment (mbarekh) [inline viewer/src/gojsTemplates.ts]: gojs documentation uses this naming http://gojs.net/latest/intro/groups.html

**Comment by nicolaserny** [comment]
@mbarekh try to check project builds on your computer (before pushing) in order to avoid multiple commits about tslint and test errors...

File: viewer/src/gojsTemplates.ts (hunk 21 above)
Comment (nicolaserny) [inline viewer/src/gojsTemplates.ts]: +1

File: viewer/src/modelTypes.ts [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+    ToId: string;
+    Type: string;
+}
+
+export interface TransitionPatternStates {
+    TransitionPatternStateData: Array<TransitionPatternStateData>;
+}
+
+export interface TransitionPatternStateData {
+    $: TransitionPatternStateDataProperties;
```
Comment (maurelio1234) [inline viewer/src/modelTypes.ts]: same comment as @nicolaserny

File: viewer/test/parser.test.ts [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+        parser = new ComponentModelParser({ model, graphical });
+
+        parser.parse().then((data: DrawComponentData) => {
+            expect(data.nodeDataArray).toEqual(correctNodeDataArray);
+            expect(data.linkDataArray).toEqual(correctLinkDataArray);
+            done();
+        });
+    });
+
+    test("Should throw an error while parsing", (done) => {
```
Comment (maurelio1234) [inline viewer/test/parser.test.ts]: while parsing invalid model files.

File: viewer/test/parser.test.ts [hunk 24]
```diff
@@ -0,0 +1,34 @@
+import * as expect from "expect";
+import { ComponentModelParser } from "../src/componentModelParser";
+import * as fs from "fs-extra";
+import { correctNodeDataArray, correctLinkDataArray } from "./parser.expectation";
+import {DrawComponentData } from "../src/gojsTemplates";
+
+describe("Test Parser", () => {
```
Comment (maurelio1234) [inline viewer/test/parser.test.ts]: should we test the case when there's no graphical file?

File: viewer/test/parser.test.ts (hunk 23 above)
Comment (maurelio1234) [inline viewer/test/parser.test.ts]: should we test for errors in the graphical model?
should we test them separetedly?

File: viewer/test/parser.test.ts (hunk 24 above)
Comment (mbarekh) [inline viewer/test/parser.test.ts]: It is in another PR

**Pull request merged**